### DISCUSSION TITLE
`depth_to_pcd` function

### DISCUSSION THREAD

**Discussion by wujiajin1021** [StanfordVL/BEHAVIOR-1K#2006] [Q&A]
it seems like there are some pointcloud accuracy error between directly get obs["pointcloud"] and depth_to_pcd function. And it seems like directly get pointcloud observation is better than depth_to_pcd
My expectation is they should have the same data, does anybody know why? Here is my code:

```import omnigibson.lazy as lazy
import omnigibson as og
import torch as th
import omnigibson.utils.transform_utils as T
from omnigibson.learning.utils.obs_utils import depth_to_pcd


# Minimal config: scene, one object, one robot, dummy task
cfg = dict()
cfg["scene"] = {"type": "Scene", "floor_plane_visible": True}
cfg["objects"] = [
    {
        "type": "PrimitiveObject",
        "name": "incredible_box",
        "primitive_type": "Cube",
        "rgba": [0, 1.0, 1.0, 1.0],
        "scale": [5, 0.5, 10],
        "fixed_base": True,
        "position": [1.0, 0, 1.0],
        "orientation": [0, 0, 0.707, 0.707],
    }
]
cfg["robots"] = [
    {
        "type": "R1Pro",
        "name": "baby_robot",
        "include_sensor_names": ["zed_link"],
        "obs_modalities": ["pointcloud", "depth_linear"],
        "sensor_config": {
            "VisionSensor": {"sensor_kwargs": {"image_height": 64, "image_width": 64}}
        },
    }
]
cfg["task"] = {
    "type": "DummyTask",
    "termination_config": dict(),
    "reward_config": dict(),
}


env = og.Environment(cfg)

# Acquire debug draw (Isaac Sim) and loop to step env and draw two pointclouds:
# - direct pointcloud (from obs) colored red
# - reconstructed depth->pcd (via depth_to_pcd) colored green
draw = lazy.isaacsim.util.debug_draw._debug_draw.acquire_debug_draw_interface()

cam_key = "baby_robot:zed_link:Camera:0"

while True:
    action = {k: v * 0 for k, v in env.action_space.sample().items()}
    obs, rew, terminated, truncated, info = env.step(action)

    # Get sensor data
    pc = obs["baby_robot"][cam_key]["pointcloud"]
    depth = obs["baby_robot"][cam_key]["depth_linear"]

    # Direct pointcloud: observations are in base frame [R,G,B,X,Y,Z]; convert to world coords
    pc_tensor = th.as_tensor(pc, dtype=th.float32)
    positions_base = pc_tensor[:, 3:6]
    valid_points = th.any(positions_base != 0.0, dim=1)
    positions_base = positions_base[valid_points]

    # Transform base-frame points to world frame using robot pose
    robot = env.robots[0]
    robot_pos, robot_quat = robot.get_position_orientation()
    robot_pos = th.as_tensor(robot_pos, dtype=th.float32, device=positions_base.device)
    robot_quat = th.as_tensor(robot_quat, dtype=th.float32, device=positions_base.device)
    robot_rot = T.quat2mat(robot_quat)
    positions_world = positions_base @ robot_rot.T + robot_pos.unsqueeze(0)

    zed_sensor = robot.sensors[f"{cam_key}"]
    K = zed_sensor.intrinsic_matrix.float()
    cam_pos_w, cam_quat_w = zed_sensor.get_position_orientation()
    cam_pos_w = th.as_tensor(cam_pos_w, dtype=th.float32)
    cam_quat_w = th.as_tensor(cam_quat_w, dtype=th.float32)
    rel_pos, rel_quat = T.relative_pose_transform(
        pos1=cam_pos_w, quat1=cam_quat_w, pos0=robot_pos, quat0=robot_quat
    )
    rel_pose = th.cat([rel_pos, rel_quat], dim=0).float().unsqueeze(0)
    depth_tensor = th.as_tensor(depth, dtype=th.float32).unsqueeze(0)

    pcd_from_depth_base = depth_to_pcd(depth=depth_tensor, rel_pose=rel_pose, K=K)[
        0
    ].reshape(-1, 3)
    pc_b_world = pcd_from_depth_base @ robot_rot.T + robot_pos.unsqueeze(0)

    # Visualization: clear and draw
    if hasattr(draw, "clear_points"):
        draw.clear_points()

    direct_pts = [
        (float(x), float(y), float(z)) for x, y, z in positions_world.cpu().numpy()
    ]
    depth_pts = [(float(x), float(y), float(z)) for x, y, z in pc_b_world.cpu().numpy()]

    if hasattr(draw, "draw_points"):
        if direct_pts:
            draw.draw_points(
                direct_pts,
                [(1.0, 0.0, 0.0, 1.0)] * len(direct_pts),
                [10.0] * len(direct_pts),
            )
        if depth_pts:
            draw.draw_points(depth_pts, [(0.0, 1.0, 0.0, 1.0)] * len(depth_pts), [10.0] * len(depth_pts))
```

<img width="1460" height="955" alt="image" src="https://github.com/user-attachments/assets/40b02258-f30b-4d8e-a0c0-ef14d1d45e67" />

### red is directly get pointcloud, green is depth_to_pcd
